Floofbot/Services/UserRoleRetentionService.cs

```diff
@@ -0,0 +1,95 @@
+﻿using Discord;
+using Discord.WebSocket;
+using Floofbot.Services.Repository;
+using Floofbot.Services.Repository.Models;
+using Microsoft.EntityFrameworkCore.Internal;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Floofbot.Services
+{
+    class UserRoleRetentionService
+    {
+        FloofDataContext _floofDb;
+        public UserRoleRetentionService(FloofDataContext floofDb)
+        {
+            _floofDb = floofDb;
+        }
+        private SocketRole GetRole(string roleId, IGuild guild)
+        {
+            foreach (SocketRole r in guild.Roles)
+            {
+                if (r.Id.ToString() == roleId)
+                    return r;
+            }
+            return null;
+        }
+        public async Task LogUserRoles(IGuildUser user)
+        {
+            string userRoles = string.Join(",", user.RoleIds);
+            await RemoveUserRoleLog(user);
+
+            _floofDb.UserRolesLists.Add(new UserRolesList{
+                ListOfRoleIds = userRoles,
+                ServerId = user.Guild.Id,
+                UserID = user.Id
+            });
+            await _floofDb.SaveChangesAsync();
+        }
+        public async Task RemoveUserRoleLog(IGuildUser user)
+        {
+            // clear old roles in db with new ones if they exist
+            if (_floofDb.UserRolesLists.AsQueryable().Where(x => x.ServerId == user.Guild.Id && x.UserID == user.Id).Any())
+            {
+                var oldUserRoles = _floofDb.UserRolesLists.AsQueryable().Where(x => x.ServerId == user.Guild.Id && x.UserID == user.Id).FirstOrDefault();
+                _floofDb.UserRolesLists.Remove(oldUserRoles);
+                await _floofDb.SaveChangesAsync();
+            }
+        }
+        public string GetUserRoles(IGuildUser user)
+        {
+            if (_floofDb.UserRolesLists.AsQueryable().Where(x => x.ServerId == user.Guild.Id && x.UserID == user.Id).Any())
+            {
+                var userRoles = _floofDb.UserRolesLists.AsQueryable().Where(x => x.ServerId == user.Guild.Id && x.UserID == user.Id).FirstOrDefault();
+                return userRoles.ListOfRoleIds;
+            }
+            else
+            {
+                return null;
+            }
+        }
+        public async Task RestoreUserRoles(IGuildUser user)
+        {
+            var oldUserRoles = GetUserRoles(user);
+
+            if (oldUserRoles != null) // user actually had old roles
+            {
+                var oldUserRolesList = oldUserRoles.Split(",");
+
+                // restore each role 1 by 1
+                foreach (string roleId in oldUserRolesList)
```

Comment:
I wonder if there should be a check for roles that _shouldn't_ be reassigned. For example, users that leave and join an excessive amount of times usually end up having to wait a week before they can be re-verified.

Reassigning roles makes sense for stuff like the muted and user roles (and maybe the moderator role), but other roles should probably be left alone IMO.